### DISCUSSION TITLE
Contact modulue changes

### DISCUSSION THREAD

**Discussion by srinath-chakravarthy** [idaholab/moose#26360] [Q&A General]
Hello all, 
Its been several months since i updated moose and my current input file for using contact is 
 [./normal_lm]
    type = NormalNodalLMMechanicalContact
    slave = 15
    master = 16
    variable = normal_lm
    master_variable = ux
    normal_smoothing_distance = 0.2
    disp_y = uy
    ncp_function_type = min
    tangential_tolerance = 0.2
    c = 1e-3
    # c = 1e-3
    # extra_vector_tags = 'ref'
    # c = 1e3
    # ncp_function_type = min
    # order = FIRST
    # tangential_tolerance = 0.1
  [../]
  [./normal_x]
    type = NormalMortarMechanicalContact
    master_boundary = 16
    slave_boundary = 15
    master_subdomain = 4
    slave_subdomain = 3
    variable = normal_lm
    slave_variable = ux
    component = x
    use_displaced_mesh = true
    compute_lm_residuals = false
    # extra_vector_tags = 'ref'
  [../]
  [./normal_y]
    type = NormalMortarMechanicalContact
    master_boundary = 16
    slave_boundary = 15
    master_subdomain = 4
    slave_subdomain = 3
    variable = normal_lm
    slave_variable = uy
    component = y
    use_displaced_mesh = true
    compute_lm_residuals = false
    # extra_vector_tags = 'ref'
  [../]

Can anyone comment on how this has changed... 

Cheers
Srinath

**Comment by GiudGiud** [comment]
Hello

Are you hitting an error? Are the results different now?

Guillaume

**Comment by srinath-chakravarthy** [comment]
i am getting an error 

*** ERROR ***
A 'NormalNodalLMMechanicalContact' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

**Comment by srinath-chakravarthy** [comment]
PS i updated moose recently and have adapted all custom kernels, materials etc so my code compiles and passes all local tests, but this was an object that used to be present in framework/contact module... looks like its been renamed

**Reply by GiudGiud** [reply]
If you look at the index in the contact module page on the framework website can you find the new name?

@recuero otherwise will know

**Reply by recuero** [reply]
That object was removed a while ago (maybe a year ago?) because we replaced the existing mortar constraints with a new formulation. If you look at the constraint folder of the contact module, you'll see similar object names without the "Nodal" string that can be used to apply surface traction on the boundaries. 

Do you need to use a traction-applying  object like that or the contact action would be sufficient for your application?

**Comment by srinath-chakravarthy** [comment]
Thanks for the response. I have several other constraint objects for custom mortar constraints i have written, for which the contact action is not exactly convenient. This type of explicit enumeration of objects is a little easier for initial proof-of-concept phase. I plan to eventually add to contact action later on. 

So the new structure will involve some kind of user object and one of the ComputeWeightedGap constraints right ?

**Reply by recuero** [reply]
Yes, you can use the user object design to develop your own mortar constraints. Then separate constraint objects are used to 1) Enforce the constraints and 2) Apply the surface tractions. 

If you already have mortar constraint objects with weighted quantities, you might only need to apply the resulting traction.

**Reply by recuero** [reply]
The idea is that the user object computes quantities over the mortar segment mesh that are then used to impose the constraints, either as penalty or with Lagrange multipliers.
